### ISSUE TITLE
Flaky Test: TestDialWaitsForServerSettingsAndFails

### ISSUE THREAD

**Issue by arvindbr8** [grpc/grpc-go#6292]
Saw this flake recently on a presubmit run. A re-run made it go away. Opening this issue to track if this comes up again.

```
--- FAIL: Test (9.54s)
    --- FAIL: Test/DialWaitsForServerSettingsAndFails (1.00s)
        tlogger.go:116: INFO clientconn.go:177 [core] [Channel #276] Channel created  (t=+186.702µs)
        tlogger.go:116: INFO clientconn.go:1628 [core] [Channel #276] original dial target is: "127.0.0.1:40827"  (t=+230.303µs)
        tlogger.go:116: INFO clientconn.go:1633 [core] [Channel #276] dial target "127.0.0.1:40827" parse failed: parse "127.0.0.1:40827": first path segment in URL cannot contain colon  (t=+250.603µs)
        tlogger.go:116: INFO clientconn.go:1649 [core] [Channel #276] fallback to scheme "passthrough"  (t=+263.903µs)
        tlogger.go:116: INFO clientconn.go:1657 [core] [Channel #276] parsed dial target is: {Scheme:passthrough Authority: URL:{Scheme:passthrough Opaque: User: Host: Path:/127.0.0.1:40827 RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}}  (t=+286.404µs)
        tlogger.go:116: INFO clientconn.go:1735 [core] [Channel #276] Channel authority set to "127.0.0.1:40827"  (t=+301.804µs)
        tlogger.go:116: INFO resolver_conn_wrapper.go:192 [core] [Channel #276] Resolver state updated: {
              "Addresses": [
                {
                  "Addr": "127.0.0.1:40827",
                  "ServerName": "",
                  "Attributes": null,
                  "BalancerAttributes": null,
                  "Type": 0,
                  "Metadata": null
                }
              ],
              "ServiceConfig": null,
              "Attributes": null
            } (resolver returned new addresses)  (t=+353.805µs)
        tlogger.go:116: INFO balancer_conn_wrappers.go:165 [core] [Channel #276] Channel switches to new LB policy "pick_first"  (t=+392.005µs)
        tlogger.go:116: INFO clientconn.go:760 [core] [Channel #276 SubChannel #277] Subchannel created  (t=+409.806µs)
        tlogger.go:116: INFO clientconn.go:451 [core] [Channel #276] Channel Connectivity change to CONNECTING  (t=+435.806µs)
        tlogger.go:116: INFO clientconn.go:1146 [core] [Channel #276 SubChannel #277] Subchannel Connectivity change to CONNECTING  (t=+458.306µs)
        tlogger.go:116: INFO clientconn.go:1260 [core] [Channel #276 SubChannel #277] Subchannel picks a new address "127.0.0.1:40827" to connect  (t=+469.307µs)
        tlogger.go:116: WARNING clientconn.go:1321 [core] [Channel #276 SubChannel #277] grpc: addrConn.createTransport failed to connect to {Addr: "127.0.0.1:40827", ServerName: "127.0.0.1:40827", }. Err: connection error: desc = "error reading server preface: read tcp 127.0.0.1:3[34](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:35)84->127.0.0.1:40827: use of closed network connection"  (t=+251.[35](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:36)0831ms)
        tlogger.go:116: INFO clientconn.go:1148 [core] [Channel #276 SubChannel #277] Subchannel Connectivity change to TRANSIENT_FAILURE, last error: connection error: desc = "error reading server preface: read tcp 127.0.0.1:33484->127.0.0.1:40827: use of closed network connection"  (t=+251.590135ms)
        tlogger.go:116: INFO clientconn.go:451 [core] [Channel #276] Channel Connectivity change to TRANSIENT_FAILURE  (t=+251.9[38](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:39)44ms)
        tlogger.go:116: INFO clientconn.go:1148 [core] [Channel #276 SubChannel #277] Subchannel Connectivity change to IDLE, last error: connection error: desc = "error reading server preface: read tcp 127.0.0.1:33484->127.0.0.1:[40](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:41)827: use of closed network connection"  (t=+252.331146ms)
        tlogger.go:116: INFO clientconn.go:[45](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:46)1 [core] [Channel #276] Channel Connectivity change to IDLE  (t=+252.[46](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:47)31[47](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:48)ms)
        tlogger.go:116: INFO clientconn.go:451 [core] [Channel #276] Channel Connectivity change to SHUTDOWN  (t=+1.00091732s)
        tlogger.go:116: INFO clientconn.go:1146 [core] [Channel #276 SubChannel #277] Subchannel Connectivity change to SHUTDOWN  (t=+1.00097332s)
        tlogger.go:116: INFO clientconn.go:1[49](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:50)6 [core] [Channel #276 SubChannel #277] Subchannel deleted  (t=+1.000994621s)
        tlogger.go:116: INFO clientconn.go:1096 [core] [Channel #276] Channel deleted  (t=+1.001006121s)
        clientconn_test.go:2[53](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191#step:8:54): dial attempts: 1; want > 1
FAIL
```

run:
https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191

**Comment by dfawley** [comment]
This is almost definitely caused by the bug @easwars found in the loop that `WithBlock` uses to wait for the connection to become ready.  It's fixed in #6263

**Comment by dfawley** [comment]
#6263 was merged so this should be good now.
